src/metabase/upload.clj

```diff
@@ -601,10 +602,84 @@
+  [table]
+  ;; For now anyone that can update a table is allowed to delete it.
+  (when-let [error (can-delete-error table)]
+    (throw error)))
+
+(defn- table-exists?
+  "Check whether the give table exists within the customer database."
+  [driver database table]
+  (try
+    (driver/describe-table driver (:id database) table)
+    true
+    (catch Exception e
+      (if (re-find #"not found" (ex-message e))
+        false
+        (throw e)))))
+
 ;;; +--------------------------------------------------
 ;;; |  public interface for updating an uploaded table
 ;;; +--------------------------------------------------
 
+(defn delete-upload!
+  "Delete the given table from both the app-db and the customer database."
+  [table & {:keys [archive-cards?]}]
+  (let [database   (table/database table)
+        driver     (driver.u/database->driver database)
+        table-name (table-identifier table)]
+    (check-can-delete table)
+
+    (try
+      ;; Attempt to delete the underlying data from the customer database.
+      ;; We perform this before marking the table as inactive in the app db so that even if it false, the table is still
+      ;; visible to administrators and the operation is easy to retry again later.
+      (driver/drop-table! driver (:id database) table-name)
+
+      ;; We mark the table as inactive synchronously, so that it will no longer shows up in the admin list.
+      (t2/update! :model/Table :id (:id table) {:active false})
+
+      ;; Ideally we would immediately trigger any further clean-up associated with the table being deactivated, but at
+      ;; the time of writing this sync isn't wired up to do anything with explicitly inactive tables, and rather
+      ;; relies on their absence from the tables being described during the database sync itself.
+      ;; TODO update the [[metabase.sync]] module to support direct per-table clean-up
+      ;; Ideally this will also clean up more the metadata which we had created around it, e.g. advanced field values.
+      #_(future (sync/retire-table! (assoc table :active false)))
+
+      ;; Archive the related cards if the customer opted in.
+      ;;
+      ;; For now, this only covers instances where the card has this as its "primary table", i.e.
+      ;; 1. A MBQL question or model that has this table as their first or only data source, or
+      ;; 2. A MBQL question or model that depends on such a model as their first or only data source.
+      ;; Note that this does not include cases where we join to this table, or even native queries which depend .
+
+      (when archive-cards?
+        (t2/update-returning-pks! :model/Card
+                                  {:table_id (:id table) :archived false}
+                                  {:archived true}))
+
+      (catch Exception e
+        (u/ignore-exceptions
+         (when (table-exists? driver database table)
+           (log/errorf e "Failure to delete table %s" (:name table))))
+        (throw e)))
```

Comment:
I don't know if we _need_ this exception handling. Exceptions while dropping a table should be rare enough and the consequences are not so dire that we need to log this case specifically. Because it's so rare there's not much benefit to this, I don't think the extra code is worth it.